### PR TITLE
Improve e2e tests metric coverage

### PR DESCRIPTION
**Pull request by machadovilaca**
### What this PR does
Before this PR:

Some metrics' existence was not ensured in e2e tests.
After the monitoring refactor it would be beneficial to have a way to add security that metrics are still being created.
 
After this PR:

Additional metric coverage in e2e tests.

jira-ticket: https://issues.redhat.com/browse/CNV-36127

Fixes #

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: 

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note

```release-note
NONE
```

### CODE REVIEW

File: tests/libmonitoring/prometheus.go [hunk 1]
```diff
@@ -82,6 +82,16 @@ func WaitForMetricValueWithLabels(client kubecli.KubevirtClient, metric string,
...
 }
 
+func WaitForMetricValueWithLabelsToBe(client kubecli.KubevirtClient, metric string, labels map[string]string, offset int, comparator string, expectedValue float64) {
+	EventuallyWithOffset(offset, func() float64 {
+		i, err := GetMetricValueWithLabels(client, metric, labels)
+		if err != nil {
+			return -1
+		}
+		return i
+	}, 3*time.Minute, 1*time.Second).Should(BeNumerically(comparator, expectedValue))
```
Comment (avlitman) [inline tests/libmonitoring/prometheus.go:92]: @machadovilaca small question: why 3 minutes?

Comment (machadovilaca) [reply tests/libmonitoring/prometheus.go:92]: same value we used before, basically:

- 1min for controllers updates
- Prometheus scrapes more or less every 30s, so +1min minimum in case it just scrapped
- +1min for margin

not too low, not unnecessarily high

File: tests/monitoring/metrics.go [hunk 2]
```diff
@@ -144,6 +125,7 @@ func fetchPrometheusMetrics(virtClient kubecli.KubevirtClient) *libmonitoring.Qu
 func setupVM(virtClient kubecli.KubevirtClient) {
 	vm := createRunningVM(virtClient)
 	libmonitoring.WaitForMetricValue(virtClient, "kubevirt_number_of_vms", 1)
+	libmonitoring.WaitForMetricValueWithLabelsToBe(virtClient, "kubevirt_vmi_filesystem_capacity_bytes", map[string]string{"namespace": vm.Namespace, "name": vm.Name}, 0, ">", 0)
```
Comment (assafad) [inline tests/monitoring/metrics.go:128]: This function is not self-explainable and this line makes it worst. It is not clear why we wait for specific metric values and why we delete the VM in this function. 
Can you please separate the test to 2 parts, one for metrics that are based on existing VM, and another for metrics that are based on a VM deletion? Then, I think we can just wait for the metrics existence in the test itself.

File: tests/monitoring/vm_monitoring.go [hunk 3]
```diff
@@ -81,6 +83,25 @@ var _ = Describe("[Serial][sig-monitoring]VM Monitoring", Serial, decorators.Sig
...
 	})
 
+	Context("VMI metrics", func() {
+		It("should have kubevirt_vmi_phase_transition_time_seconds buckets correctly configured", func() {
+			vmi := libvmifact.NewGuestless()
+			tests.RunVMIAndExpectLaunch(vmi, 240)
+
+			buckets := virtcontroller.PhaseTransitionTimeBuckets()
+			labels := map[string]string{"le": strconv.FormatFloat(buckets[0], 'f', -1, 64)}
+			libmonitoring.WaitForMetricValueWithLabelsToBe(virtClient, "kubevirt_vmi_phase_transition_time_seconds_bucket", labels, 0, ">", 0)
```
Comment (assafad) [inline tests/monitoring/vm_monitoring.go:93]: In the old location we checked for all buckets existence, why do you only the first now?

File: tests/monitoring/vm_monitoring.go [hunk 4]
```diff
@@ -81,6 +83,25 @@ var _ = Describe("[Serial][sig-monitoring]VM Monitoring", Serial, decorators.Sig
...
+		})
+
+		It("should have kubevirt_rest_client_requests_total for the 'virtualmachineinstances' resource", func() {
+			vmi := libvmifact.NewGuestless()
+			tests.RunVMIAndExpectLaunch(vmi, 240)
+
+			labels := map[string]string{"resource": "virtualmachineinstances"}
+			libmonitoring.WaitForMetricValueWithLabelsToBe(virtClient, "kubevirt_rest_client_requests_total", labels, 0, ">", 0)
```
Comment (assafad) [inline tests/monitoring/vm_monitoring.go:101]: same

Comment (machadovilaca) [reply tests/monitoring/vm_monitoring.go:101]: this is logically the same as before, no difference

File: tests/monitoring/vm_monitoring.go (hunk 3 above)
Comment (machadovilaca) [reply tests/monitoring/vm_monitoring.go:93]: this was flaky because not always all the histograms buckets were populated, but we know that having 1 VMI, at least the lowest bucket must have a value > 0

File: tests/monitoring/metrics.go (hunk 2 above)
Comment (machadovilaca) [reply tests/monitoring/metrics.go:128]: we would make the test take almost twice as long, maybe we can just refactor the function name and maybe extract some logic to an helper function

File: tests/libmonitoring/prometheus.go (hunk 1 above)
Comment (avlitman) [reply tests/libmonitoring/prometheus.go:92]: @machadovilaca thanks!

File: tests/monitoring/vm_monitoring.go (hunk 3 above)
Comment (assafad) [reply tests/monitoring/vm_monitoring.go:93]: > not always all the histograms buckets were populated

How so? Buckets are lower-equal, no?

File: tests/monitoring/metrics.go (hunk 2 above)
Comment (assafad) [reply tests/monitoring/metrics.go:128]: Why would that make the test so long?

File: tests/monitoring/vm_monitoring.go (hunk 3 above)
Comment (machadovilaca) [reply tests/monitoring/vm_monitoring.go:93]: +Inf would have the total count? let me try...

File: tests/monitoring/metrics.go (hunk 2 above)
Comment (assafad) [reply tests/monitoring/metrics.go:128]: Now you propose to:
create -> check some metrics -> delete -> check one metric -> check all the metrics
I propose:
create -> check all metrics expect of deletion based ones -> delete -> check deletion based metrics

Can it work? This way we only check the metrics in the test itself and separate the VMs work from the metrics.

Comment (assafad) [reply tests/monitoring/metrics.go:128]: ok, let's try to refactor the function, to make it clearer.

Comment (machadovilaca) [reply tests/monitoring/metrics.go]: updated @assafad, let me know what you think

File: tests/monitoring/metrics.go [hunk 5]
```diff
@@ -141,19 +122,16 @@ func fetchPrometheusMetrics(virtClient kubecli.KubevirtClient) *libmonitoring.Qu
...
-	err := virtClient.VirtualMachine(vm.Namespace).Delete(context.Background(), vm.Name, metav1.DeleteOptions{})
-	Expect(err).ToNot(HaveOccurred())
-
-	libmonitoring.WaitForMetricValue(virtClient, "kubevirt_number_of_vms", -1)
+func basicVMLifecycle(virtClient kubecli.KubevirtClient) {
+	vm := createAndRunVM(virtClient)
+	waitForRunningVMMetrics(virtClient, vm)
+	deleteVM(virtClient, vm)
+	waitForDeletedVMMetrics(virtClient, vm)
 }
```
Comment (assafad) [inline tests/monitoring/metrics.go:130]: I feel we can remove this func and do it from the BeforeEach(), wdyt?

Comment (machadovilaca) [reply tests/monitoring/metrics.go:130]: I think in this use case it makes sense to be here. What would be the benefits of moving?

Comment (assafad) [reply tests/monitoring/metrics.go:130]: Removing one level from the code's depth. This function is used only once and it ONLY calls other functions. Feels redundant.

Comment (assafad) [reply tests/monitoring/metrics.go:130]: But this is a nit. If you disagree let me know and I'll LGTM.

Comment (machadovilaca) [reply tests/monitoring/metrics.go:130]: it can be inlined in this function, moving upwards I don't think so

Comment (machadovilaca) [reply tests/monitoring/metrics.go:130]: but then we would go back to the first implementation

**Comment by assafad** [comment]
thanks @machadovilaca 
/lgtm

File: tests/monitoring/vm_monitoring.go [hunk 6]
```diff
@@ -40,6 +41,7 @@ import (
 	"kubevirt.io/client-go/kubecli"
 
 	"kubevirt.io/kubevirt/pkg/libvmi"
+	virtcontroller "kubevirt.io/kubevirt/pkg/monitoring/metrics/virt-controller"
```
Comment (enp0s3) [inline tests/monitoring/vm_monitoring.go:44]: @machadovilaca I think its better not to use product code in e2e, but implement the needed function in-place. Otherwise you are losing the protection when refactoring the product code.

File: tests/monitoring/vm_monitoring.go [hunk 7]
```diff
@@ -81,6 +83,28 @@ var _ = Describe("[Serial][sig-monitoring]VM Monitoring", Serial, decorators.Sig
...
 
+	Context("VMI metrics", func() {
+		It("should have kubevirt_vmi_phase_transition_time_seconds buckets correctly configured", func() {
+			vmi := libvmifact.NewGuestless()
+			tests.RunVMIAndExpectLaunch(vmi, 240)
+
+			for _, bucket := range virtcontroller.PhaseTransitionTimeBuckets() {
+				labels := map[string]string{"le": strconv.FormatFloat(bucket, 'f', -1, 64)}
+
+				GinkgoLogr.Info("Checking bucket", "labels", labels)
```
Comment (enp0s3) [inline tests/monitoring/vm_monitoring.go:94]: @machadovilaca Is this print necessary? we usually try to avoid verbosity in tests unless it really needed. From past experience we know that people are frustrated to analyze failed e2e tests because of plenty of printouts.

File: tests/monitoring/metrics.go [hunk 8]
```diff
@@ -141,19 +122,26 @@ func fetchPrometheusMetrics(virtClient kubecli.KubevirtClient) *libmonitoring.Qu
...
 	Expect(err).ToNot(HaveOccurred())
 
+	By("Waiting for the VM deletion to be reported")
 	libmonitoring.WaitForMetricValue(virtClient, "kubevirt_number_of_vms", -1)
 }
 
-func createRunningVM(virtClient kubecli.KubevirtClient) *v1.VirtualMachine {
-	vmi := libvmifact.NewGuestless(libvmi.WithNamespace(testsuite.GetTestNamespace(nil)))
+func createAndRunVM(virtClient kubecli.KubevirtClient) *v1.VirtualMachine {
+	vmi := libvmifact.NewFedora(libvmi.WithNamespace(testsuite.GetTestNamespace(nil)))
```
Comment (enp0s3) [inline tests/monitoring/metrics.go:144]: @machadovilaca Why did you change it to a bootable guest?

File: tests/monitoring/vm_monitoring.go (hunk 6 above)
Comment (machadovilaca) [reply tests/monitoring/vm_monitoring.go:44]: The alternative would be to hard code the bucket values, which would still mean the need to change when refactoring or even changing those values

File: tests/monitoring/metrics.go (hunk 8 above)
Comment (machadovilaca) [reply tests/monitoring/metrics.go:144]: Now testing the domainstats metrics, a bootable guest is need to ensure they are collected and with all the information we expose

File: tests/monitoring/vm_monitoring.go (hunk 6 above)
Comment (enp0s3) [reply tests/monitoring/vm_monitoring.go:44]: @machadovilaca And this is good IMO, these are the buckets you expect to see, so its fine to hardcode them in the test, so that if someone will change them the test will fail.
